homeassistant/components/yolink/sensor.py

```diff
@@ -1,24 +1,106 @@
-        if device["type"] == "DoorSensor":
-            devices.append(YoLinkDoorSensor(device, hass, config_entry))
+SENSOR_DESC_MAP: dict[str, SensorEntityDescription] = {
+    "battery": SensorEntityDescription(
+        key="battery",
+        device_class=SensorDeviceClass.BATTERY,
+        native_unit_of_measurement=PERCENTAGE,
+        name="Battery",
+        state_class=SensorStateClass.MEASUREMENT,
+    ),
+}
 
-    for device in devices:
-        entities.extend(device.entities)
+SENSOR_DEVICE_TYPE = [ATTR_DEVICE_DOOR_SENSOR]
 
-    hass.data[DOMAIN][config_entry.entry_id][HOME_SUBSCRIPTION].attachPlatformDevices(
-        Platform.SENSOR, devices
-    )
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up YoLink Sensor from a config entry."""
+    sensor_devices = [
+        device
+        for device in hass.data[DOMAIN][config_entry.entry_id][ATTR_DEVICE]
+        if device[ATTR_DEVICE_TYPE] in SENSOR_DEVICE_TYPE
+    ]
+    yl_client = hass.data[DOMAIN][config_entry.entry_id][ATTR_CLIENT]
+    yl_mqtt_client = hass.data[DOMAIN][config_entry.entry_id][ATTR_MQTT_CLIENT]
+    entities = []
+    for sensor_device in sensor_devices:
+        sensor_entities = get_sensor_entities(hass, sensor_device, yl_client)
+        if sensor_entities is not None:
+            for entity in sensor_entities:
+                entities.append(entity)
+    yl_mqtt_client.add_device_subscription(ATTR_PLATFORM_SENSOR, entities)
     async_add_entities(entities)
+
+
+def get_sensor_entities(
+    hass: HomeAssistant, device: dict, client: YoLinkClient
+) -> list[YoLinkEntity] | None:
+    """Get device entities."""
+    if device[ATTR_DEVICE_TYPE] == ATTR_DEVICE_DOOR_SENSOR:
+        return [YoLinkSensorBatteryEntity(hass, device, client)]
+    return None
+
+
+class YoLinkSensorBatteryEntity(YoLinkEntity, SensorEntity):
+    """YoLink Sensor Entity."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        device: dict,
+        client: YoLinkClient,
+    ) -> None:
+        """Init YoLink Sensor."""
+        super().__init__(hass, device, client)
+        self.entity_description = SENSOR_DESC_MAP["battery"]
+        self._attr_unique_id = f"{device['deviceId']} {self.entity_description.key}"
+        self._attr_name = f"{device['name']} ({self.entity_description.name})"
+
+    async def async_added_to_hass(self) -> None:
+        """Add to hass."""
+
+        async def request_state():
+            resp = await self.fetch_state_with_api()
+            if "state" in resp.data:
+                await self.update_entity_state(resp.data[CONF_STATE])
+
+        self.hass.create_task(request_state())
+
+    async def update_entity_state(self, state: dict) -> None:
+        """Update HA Entity State."""
+        if state is None:
+            return
+        self._attr_native_value = percentage.ordered_list_item_to_percentage(
```

Comment:
Extract the value format function to an attribute on the entity description and use this when setting the `native_value`.

Example:
https://github.com/home-assistant/core/blob/bbe807c655edcfe2c8b354dd3685221e4a2755a7/homeassistant/components/airly/sensor.py#L59-L63

https://github.com/home-assistant/core/blob/bbe807c655edcfe2c8b354dd3685221e4a2755a7/homeassistant/components/airly/sensor.py#L115

https://github.com/home-assistant/core/blob/bbe807c655edcfe2c8b354dd3685221e4a2755a7/homeassistant/components/airly/sensor.py#L166-L170